flips/sync-protocol.md

```diff
@@ -0,0 +1,90 @@
+* If the response is for a Batch Request, it should contain a subset of the requested block IDs.
+
+Since the nonce allows us to distinguish between responses to Batch and Range Requests, we can avoid optimistically setting the statuses of heights as Received for responses to Batch Requests.
+
+At any time, there is a single range of heights that the Sync Engine actively requests, which is tracked by the Sync Core. We call this the Active Range. The Active Range is parameterized by two variables, which effectively replace the `Heights` map from the existing implementation:
+* `RangeStart` is the start height of the range, which is greater than or equal to `LocalFinalizedHeight`, the local finalized block height.
+* `RangeEnd` is the end height of the range, which is less than or equal to `TargetFinalizedHeight`, the speculated finalized block height of the overall chain (more details below).
+
+The Active Range has a default size of `DefaultRangeSize`. `RangeStart` is updated according to the following conditions, and `RangeEnd` should always be updated as necessary to satisfy `RangeEnd - RangeStart >= DefaultRangeSize`:
+* If the local finalized height has increased by `h`, increase `RangeStart` by `h`.
+* If every height in the Active Range has been received from at least `MinResponses` different nodes, increase `RangeStart` by `DefaultRangeSize`.
+
+> **Note:** The second condition is needed to cover the last item in [Potential Problems](#potential-problems). To guarantee liveness, `MinResponses` should be set to a value greater than the maximum tolerable number of Byzantine nodes. Alternatively, we could increase `RangeEnd` by `DefaultRangeSize` instead of increasing `RangeStart`, thereby allowing the Active Range to grow without bound and obviating the need for any lower bound on `MinResponses`. 
+> 
+> A possible optimization here is that instead of waiting for the entire Active Range to satisfy the update condition, we can update as soon as any sub-range starting from `RangeStart` satisfies the condition.
+
+The Active Range can be broken up and requested by the Sync Engine in multiple segments, similar to the existing implementation.
+
+We track the list of responses we've received for each height in a data structure called `Responses`, which we can use to avoid sending duplicate requests to the same node. We can also use it later on to slash nodes for sending faulty responses. 
+
+The value of `TargetFinalizedHeight` should reflect the Sync Height Responses that have been received, yet account for the possibility that some of these responses are malicious. Therefore, the Sync Height Response processing logic should incorporate some sort of expiration mechanism. Some possibilities for determining `TargetFinalizedHeight` are:
+* Calculate a moving average of all the Sync Height Responses received. 
```

Comment:
Idea: make `SyncResponse`s commit to a block ID corresponding to their claimed latest finalized height, for slashing.

Also, I think significant outliers would pose a problem to all 3 of these approaches. We should consider discarding responses which are obviously invalid (eg. days or weeks ahead of the mean).

Reply:
@jordanschalm That's a good idea. As for outliers, I think this can mostly be mitigated by taking the median instead of the mean.